pouta_blueprints/drivers/provisioning/openshift_driver.py

```diff
@@ -0,0 +1,638 @@
+                current_ts = int(time.time())
+            if self.token_data['expires_at'] - 600 < current_ts:
+                self.token_data = self._request_token(current_ts)
+
+        return self.token_data['access_token']
+
+    def _construct_url(self, kubeapi, namespace=None, object_kind=None, object_id=None):
+        if kubeapi:
+            url_components = [self.kube_base_url]
+        else:
+            url_components = [self.oapi_base_url]
+
+        if namespace:
+            url_components.append('namespaces')
+            url_components.append(namespace)
+        if object_kind:
+            url_components.append(object_kind)
+        if object_id:
+            url_components.append(object_id)
+        url = '/'.join(url_components)
+
+        return url
+
+    def make_request(self, method=None, kubeapi=False, verbose=False, namespace=None, object_kind=None, object_id=None,
+                     params=None, data=None, raise_on_failure=True):
+        url = self._construct_url(kubeapi, namespace, object_kind, object_id)
+        headers = {'Authorization': 'Bearer %s' % self._get_token()}
+        if isinstance(data, dict):
+            data = json.dumps(data)
+
+        if data:
+            if not method or method == 'POST':
+                resp = self._session.post(url, headers=headers, verify=False, params=params, data=data)
+            elif method == 'PUT':
+                resp = self._session.put(url, headers=headers, verify=False, params=params, data=data)
+            else:
+                raise RuntimeError('Do not know what to do with data and method %s' % method)
+        else:
+            if method and method != 'GET':
+                raise RuntimeError('Do not know what to do with no data and method %s' % method)
+            resp = self._session.get(url, headers=headers, verify=False, params=params)
+        if verbose:
+            self.print_resp(resp)
+
+        if raise_on_failure and not resp.ok:
+            raise RuntimeError(resp.text)
+
+        return resp
+
+    def make_delete_request(self, kubeapi=False, verbose=False, namespace=None, object_kind=None, object_id=None,
+                            raise_on_failure=True):
+        url = self._construct_url(kubeapi, namespace, object_kind, object_id)
+        headers = {'Authorization': 'Bearer %s' % self._get_token()}
+
+        resp = self._session.delete(url, headers=headers, verify=False)
+        if verbose:
+            self.print_resp(resp)
+
+        if raise_on_failure and not resp.ok:
+            raise RuntimeError(resp.text)
+
+        return resp
+
+    def search_by_label(self, kubeapi=False, namespace=None, object_kind=None, params=None):
+        res = self.make_request(
+            kubeapi=kubeapi,
+            namespace=namespace,
+            object_kind=object_kind,
+            params=params
+        )
+        res_json = res.json()
+        return res_json.get('items', [])
+
+
+class OpenShiftDriverAccessProxy(object):
```

Comment:
The proxy is there to abstract all external access away from the driver itself, so it is easier to build a mock object for unit tests.